### DISCUSSION TITLE
Fine-tuning: Energy Offset Issue

### DISCUSSION THREAD

**Discussion by sek1ro-yuzzz** [MDIL-SNU/SevenNet#218] [Q&A] [answered]
Hello everyone,
I'm currently fine-tuning a SevenNet model for solid electrolyte NaTaCl6 systems. I've generated a dataset of approximately 3000 structures from MD trajectories with DFT static calculations for the fine-tuning process. I choose 7net-mf-ompa (mpa) as pre-trained model.
**Results**
The fine-tuned model performs exceptionally well for force predictions - I'm very impressed with the accuracy! 
However, I've noticed a systematic offset in energy predictions, which I've seen mentioned in other discussions as well.I understand from those discussions that increasing the number of epochs (although my training seems to converge around 40-50 epochs already) or reducing the force loss weight might help address this issue.
**Configuration Settings**
I'd like to ask whether my current settings are reasonable for this application:
```
config.update({
    'optimizer': 'adam',
    'optim_param': {'lr': 0.001},
    'scheduler': 'exponentiallr',     
    'scheduler_param': {'gamma': 0.99},
    'is_ddp': False,
    'force_loss_weight': 1.0,
    'stress_loss_weight': 0.01,
    'loss': 'Huber',
    'loss_param': {'delta': 0.01},
    'per_epoch': 10,
    'train_shift_scale': True, 
    'train_denominator': True,   
})
```
I'm particularly wondering whether I should keep train_shift_scale and train_denominator enabled. Could these settings be contributing to the energy offset issue? What would be the implications of disabling them?
![PixPin_2025-05-17_16-26-34](https://github.com/user-attachments/assets/b0349b5c-b2f6-4ec1-96c0-03342b2892db)

**Comment by sek1ro-yuzzz** [comment]
Can I finetune like this?
Step 1: set "force_loss_weight=1" and "stress_loss_weight=0.01" 
then Step 2: set "force_loss_weight=0" and "stress_loss_weight=0" to specialize in fine-tuning the energy

**Comment by YutackPark** [comment, marked as answer]
For me, the config seems perfectly fine except for `train_denominator = true`. I recommend to set it False.

On the other hand, altought there exists systematic shift, the overall error is very small (below 4 meV/atom). I haven't tried to fine-tune more if the energy error is already that small.

Systamatice 'energy' shift MLIP of common, especially for pre-trained models as the pre-training datasets DFT setting is usually not same as that of fine-tuning dataset, but it is often not a big problem as the absolute total energy is physically meaningless.

+ for second comment, I would try changing scheduler or learning rate instead of setting force weight as 0.

**Reply by sek1ro-yuzzz** [reply]
Thank you very much for the helpful feedback! I'll continue fine-tuning with the suggested adjustments.
I have an additional question after reviewing my fine-tuning logs. I noticed that both the Train and Valid Energy_RMSE are around 0.5 meV/atom, which suggests the model should be predicting energies very accurately. However, when I plot the predicted vs. DFT energies, I observe a systematic offset of approximately 3 meV/atom.
If the Energy_RMSE is so low (~0.5 meV), why does the energy parity plot show a systematic shift of ~3 meV? 
Thank you again for your time and assistance!
![image](https://github.com/user-attachments/assets/1c6c48ab-4ce8-41ca-98da-a11b7919040e)

**Comment by sek1ro-yuzzz** [comment]
I rewrote my fine-tuning code and restarted the process beginning with data preparation(and even changed my GPU card to V100, lol). Everything is working normally now! I haven't identified the exact cause yet, but it's worth mentioning that I was previously using sevenn_graph_build structure_list 5.0, and later discovered that SevenNet-MF-ompa seems to ues a cutoff of 6.0.
Here's the performance after just three epochs of fine-tuning:
![image](https://github.com/user-attachments/assets/fe38434f-c1b1-49b1-8779-6dca8611eae0)

Thank you again for your excellent work on developing SevenNet!

**Reply by YutackPark** [reply]
> sevenn_graph_build structure_list 5.0, and later discovered that SevenNet-MF-ompa seems to ues a cutoff of 6.0

That makes sense. Maybe you could find a warning in your previous log. As it is a graph, model can process data with different cutoffs. But it is highly discouraged since 1) it doesn't make sense 2) different from the training set
